pkg/xds/translate.go

```diff
@@ -40,8 +41,186 @@ const (
 	HeaderScheme    = ":scheme"
 )
 
-// ClusterName specifies cluster name for a destination
-type ClusterName func(*routingv2.Destination) string
+const (
+	// UnresolvedCluster for destinations pointing to unknown clusters.
+	UnresolvedCluster = "unresolved-cluster"
+
+	// DefaultOperation is the default decorator
+	DefaultOperation = "default-operation"
+)
+
+// ServiceByName claims a service entry from the registry using a host name.
+type ServiceByName func(host string, contextNamespace string) *model.Service
+
+// SubsetSelector resolves a subset to labels.
+type SubsetSelector func(service *model.Service, subset string) map[string]string
+
+// GuardedHost is a context-dependent virtual host entry with guarded routes.
+type GuardedHost struct {
+	// Port is the capture port (e.g. service port)
+	Port int
+
+	// Services are the services matching the virtual host.
+	// The service host names need to be contextualized by the source.
+	Services []*model.Service
+
+	// Hosts is a list of alternative literal host names for the host.
+	Hosts []string
+
+	// Routes in the virtual host
+	Routes []GuardedRoute
+}
+
+// TranslateVirtualHosts creates the entire routing table for Istio v1alpha2 configs.
+// Services are indexed by FQDN hostnames.
+// Cluster domain is used to resolve short service names (e.g. "svc.cluster.local").
+func TranslateVirtualHosts(
+	serviceConfigs []model.Config,
+	services map[string]*model.Service,
+	subsetSelector SubsetSelector,
+	clusterDomain string) []GuardedHost {
+	out := make([]GuardedHost, 0)
+
+	serviceByName := func(host string, contextNamespace string) *model.Service {
+		if strings.Contains(host, ".") {
+			return services[host]
+		}
+
+		return services[fmt.Sprintf("%s.%s.%s", host, contextNamespace, clusterDomain)]
+	}
+
+	// translate all virtual service configs
+	for _, config := range serviceConfigs {
+		out = append(out, TranslateVirtualHost(config, serviceByName, subsetSelector)...)
+	}
+
+	// compute services missing service configs
+	missing := make(map[string]bool)
+	for fqdn := range services {
+		missing[fqdn] = true
+	}
+	for _, host := range out {
+		for _, service := range host.Services {
+			delete(missing, service.Hostname)
+		}
+	}
+
+	// append default hosts for the service missing virtual services
+	for fqdn := range missing {
+		svc := services[fqdn]
+		for _, port := range svc.Ports {
+			if port.Protocol.IsHTTP() {
```

Comment:
Here and below in `TranslateVirtualHost`, why are we only generating hosts for HTTP traffic? Doesn't the same logic apply for TCP traffic at this point (i.e. we have a model.Service even for TCP services, don't we)?

Reply:
Because TCP does not have routes or RDS. Its statically created filter with almost empty config in v2.